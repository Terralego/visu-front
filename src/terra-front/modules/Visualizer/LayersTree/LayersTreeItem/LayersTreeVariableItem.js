import { Callout, Collapse } from '@blueprintjs/core';
import React, { useContext, useEffect, useState } from 'react';
import { useTranslation } from 'react-i18next';
import LayersTreeItem from '.';
import Select from '../../../Forms/Controls/Select';
import context from '../LayersTreeProvider/context';

const getSelectedLayer = (selectedVariables, layers) =>
  layers.find(layer =>
    Object.entries(selectedVariables).every(
      ([variable, value]) => layer.variables?.[variable] === value,
    ));

const getValuesByVariable = (variables, layers) => {
  const values = {};
  variables.forEach(variable => {
    values[variable.id] = Array.from(
      new Set(layers.map(layer => layer.variables?.[variable.id])),
    ).sort();
  });
  return values;
};

const LayersTreeVariableItem = ({ group, layers, activeLayer }) => {
  const { t } = useTranslation();

  const { variables } = group;

  const [selectedVariables, setSelectedVariables] = React.useState(
    layers[0].variables || {},
  );


  const [layer, setLayer] = useState(
    getSelectedLayer(selectedVariables, layers),
  );
  const [error, setError] = useState(null);

  const { setLayerState, getLayerState, layersExtent, isDetailsVisible } = useContext(context);

  const valuesByVariable = getValuesByVariable(variables, layers);

  useEffect(() => {
    if (!activeLayer) return;
    setLayer(activeLayer);
    setSelectedVariables(activeLayer.variables);
    setLayerState({ layer: activeLayer, state: { active: true } });
  }, [activeLayer, setLayerState]);

  const handleChange = (variable, value) => {
    const newSelectedVariables = { ...selectedVariables, [variable]: value };
    const newLayer = getSelectedLayer(newSelectedVariables, layers);
    setSelectedVariables(newSelectedVariables);
    if (!newLayer) {
      setError(t('terralego.visualizer.layerstree.variableLayers.error'));
      return;
    }
    setError(null);
    setLayerState({ layer, state: { active: false } });
    setLayerState({ layer: newLayer, state: { active: true } });
    setLayer(newLayer);
  };

  const layerActive = getLayerState({ layer }).active;

  return (
    <>
      <div className="layerstree-group layerstree-group--active">
        {layer && (
          <LayersTreeItem
            key={layer.label}
            layer={layer}
            customLabel={
              layerActive
                ? `${group.group} (${variables.map(({ id }) => layer.variables[id]).join(' - ')})`
                : group.group
            }
            extent={layersExtent?.[layer.label]}
            isDetailsVisible={isDetailsVisible}
          />
        )}
        <Collapse isOpen={layerActive}>
          <div style={{ padding: 5, display: 'flex', flexDirection: 'column', gap: '0.8rem' }}>
            {error && <Callout intent="danger">{error}</Callout>}
            {variables.map(variable => (
              <Select
                fullWidth
                value={selectedVariables[variable.id]}
                onChange={value => handleChange(variable.id, value)}
                values={valuesByVariable[variable.id]}
              />
            ))}
          </div>
        </Collapse>
      </div>
    </>
  );
};

export default LayersTreeVariableItem;
