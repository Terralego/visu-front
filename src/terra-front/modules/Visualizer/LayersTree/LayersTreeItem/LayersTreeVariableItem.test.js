import React, { createContext as mockCreateContext } from 'react';
import renderer from 'react-test-renderer';
import LayersTreeVariableItem from './LayersTreeVariableItem';

jest.mock('../LayersTreeProvider/context', () => {
  const context = mockCreateContext({
    setLayerState: () => {},
    getLayerState: () => {},
    layersExtent: () => {},
    isDetailsVisible: () => {},
  });
  return context;
});
jest.mock(
  '.',
  () =>
    function LayersTreeItem () {
      return <p>LayersTreeItem</p>;
    },
);

describe('LayersTreeVariableItem', () => {
  it('should render', () => {
    const layers = [
      { variable: 'variable1', value: 'value1', active: true },
      { variable: 'variable2', value: 'value2', active: false },
    ];
    const group = {
      variables: [
        { id: 'variable1', label: 'Variable 1' },
        { id: 'variable2', label: 'Variable 2' },
      ],
      group: 'Group 1',
      layers,
    };

    const tree = renderer.create(<LayersTreeVariableItem group={group} layers={layers} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
