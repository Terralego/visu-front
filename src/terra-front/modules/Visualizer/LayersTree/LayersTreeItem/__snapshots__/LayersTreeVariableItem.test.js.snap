// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LayersTreeVariableItem should render 1`] = `
<div
  className="layerstree-group layerstree-group--active"
>
  <p>
    LayersTreeItem
  </p>
  <div
    className="bp3-collapse"
    style={
      Object {
        "height": undefined,
        "overflowY": undefined,
        "transition": undefined,
      }
    }
  >
    <div
      aria-hidden={false}
      className="bp3-collapse-body"
      style={
        Object {
          "transform": "translateY(-undefinedpx)",
          "transition": undefined,
        }
      }
    />
  </div>
</div>
`;
